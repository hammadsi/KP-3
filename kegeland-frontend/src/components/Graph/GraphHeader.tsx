import { HStack } from '@chakra-ui/react';
import moment from 'moment';
import { MdOutlineWatchLater } from 'react-icons/md';
import LabeledValue from '../LabeledValue';

type GraphHeaderProps = {
  date: number;
  endTime: Date;
};

const GraphHeader: React.FC<GraphHeaderProps> = ({ date, endTime }) => {
  const dateFormat = new Date(date);
  const formattedEndTime = endTime.toLocaleString(); 

  function getDuration(startDate: number, endDate: Date) {
    const durationInMilliseconds = endDate.getTime() - startDate;

    // if duratation is less than 10 seconds, return 0:00:10
    if (durationInMilliseconds < 10000) {
      return '0:00:10';
    }

    const hours = Math.floor(durationInMilliseconds / (1000 * 60 * 60));
    const minutes = Math.floor(
      (durationInMilliseconds % (1000 * 60 * 60)) / (1000 * 60),
    );
    const seconds = Math.floor((durationInMilliseconds % (1000 * 60)) / 1000);

    return `${hours}:${String(minutes).padStart(2, '0')}:${String(
      seconds,
    ).padStart(2, '0')}`;
  }

  const duration = getDuration(date, new Date(formattedEndTime));

  return (
    <HStack justifyContent="space-between">
      <HStack spacing={12}>
        <LabeledValue
          label="Date of exercise"
          value={moment(dateFormat).format('lll')}
        />
        <LabeledValue
          label="Duration"
          icon={MdOutlineWatchLater}
          value={`${duration}`}
        />
      </HStack>
    </HStack>
  );
};

export default GraphHeader;
