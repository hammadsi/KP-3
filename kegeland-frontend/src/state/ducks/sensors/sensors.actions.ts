import { createAsyncThunk } from '@reduxjs/toolkit';

import { apiCaller } from '../../../utils/apiCaller';

import { Sensor, SensorType } from './sensors.interface';

export const fetchSensor = createAsyncThunk(
  'sensors/fetchSensor',
  async (sensorType: SensorType) =>
    apiCaller<Sensor>({
      url: `sensors/${sensorType}`,
      method: 'GET',
    }),
);
