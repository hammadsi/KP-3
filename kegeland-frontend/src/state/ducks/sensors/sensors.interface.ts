export enum SensorType {
  FEMFIT = 'femfit',
  EMPATICA = 'empatica',
  IMU = 'imu',
}

export type Sensor = {
  id: string;
  name: SensorType;
  labels: string[];
};

export interface SensorState {
  loading: boolean;
  data: {
    [key in SensorType]?: Sensor;
  };
  error: string | undefined;
}
