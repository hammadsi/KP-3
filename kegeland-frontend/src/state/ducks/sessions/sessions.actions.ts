import { createAsyncThunk } from '@reduxjs/toolkit';

import { apiCaller } from '../../../utils/apiCaller';

import { FetchSessionsDto, Session } from './sessions.interface';

export const fetchSessionById = createAsyncThunk(
  'sessions/fetchSessionById',
  async (id: string) =>
    apiCaller<Session>({
      url: `sessions/${id}`,
      method: 'GET',
    }),
);

export const fetchSessions = createAsyncThunk(
  'sessions/fetchSessions',
  async (params: FetchSessionsDto) =>
    apiCaller<Session[]>({
      url: 'sessions',
      method: 'GET',
      params,
    }),
);
