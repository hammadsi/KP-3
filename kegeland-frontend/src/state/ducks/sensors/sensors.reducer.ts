import { createSlice } from '@reduxjs/toolkit';

import {
  isPendingAction,
  isFulfilledAction,
  isRejectedAction,
} from '../../../utils/thunk.utils';

import { fetchSensor } from './sensors.actions';
import { SensorState } from './sensors.interface';

export const initialState: SensorState = {
  loading: false,
  data: {},
  error: undefined,
};

const sensorSlice = createSlice({
  name: 'sensors',
  initialState,
  reducers: {
    clearSensorsState: (state) => {
      state.loading = false;
      state.data = {};
      state.error = undefined;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchSensor.fulfilled, (state, { payload }) => {
        state.data[payload.name] = payload;
      })
      .addMatcher(
        (action) => isPendingAction(action, 'sensors'),
        (state) => {
          state.loading = true;
          state.error = undefined;
        },
      )
      .addMatcher(
        (action) => isFulfilledAction(action, 'sensors'),
        (state) => {
          state.loading = false;
          state.error = undefined;
        },
      )
      .addMatcher(
        (action) => isRejectedAction(action, 'sensors'),
        (state, { error }) => {
          state.loading = false;
          state.error = error.message;
        },
      );
  },
});

export const { clearSensorsState } = sensorSlice.actions;

export const sensorReducer = sensorSlice.reducer;
