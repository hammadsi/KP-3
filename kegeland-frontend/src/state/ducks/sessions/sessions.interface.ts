import { SensorType } from '../sensors/sensors.interface';
import { GameSession } from '../wheelchairPatients/wheelchairPatients.interface';

export type SessionDataPoint = number | null;

export type LeanSession = {
  id: string;
  userId: string;
  sensor: SensorType;
  createdAt: number;
};

export type ViewSession = LeanSession | GameSession;

export interface Session extends LeanSession {
  data: Record<string, SessionDataPoint[]>;
}

export type SessionsState = {
  loading: boolean;
  session: Session | undefined;
  data: Session[];
  error: string | undefined;
};

export type FetchSessionsDto = {
  sensor?: SensorType;
  userId?: string;
};
