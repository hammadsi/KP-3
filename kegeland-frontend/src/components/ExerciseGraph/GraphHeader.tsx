import { HStack, IconButton, Tooltip } from '@chakra-ui/react';
import { capitalize } from 'lodash';
import moment from 'moment';
import {
  MdDock,
  MdDownload,
  MdOutlineWatchLater,
  MdSettings,
} from 'react-icons/md';

import { Sensor } from '../../state/ducks/sensors/sensors.interface';
import { Session } from '../../state/ducks/sessions/sessions.interface';
import { exportSessionAsCsv } from '../../utils/csv.utils';
import { getSessionDuration } from '../../utils/graph.utils';
import LabeledValue from '../LabeledValue';

type GraphHeaderProps = {
  sensor: Sensor;
  session: Session;
  toggleSettings: () => void;
};

const GraphHeader: React.FC<GraphHeaderProps> = ({
  sensor,
  session,
  toggleSettings,
}) => {
  return (
    <HStack justifyContent="space-between">
      <HStack spacing={12}>
        <LabeledValue
          label="Date of exercise"
          value={moment(session.createdAt).format('lll')}
        />

        <LabeledValue
          label="Device"
          icon={MdDock}
          value={capitalize(sensor.name)}
        />

        <LabeledValue
          label="Duration"
          icon={MdOutlineWatchLater}
          value={`${moment
            .utc(getSessionDuration(session))
            .format('HH:mm:ss')}`}
        />
      </HStack>
      <HStack spacing={0}>
        <IconButton
          aria-label="Open graph settings"
          icon={<MdSettings />}
          variant="unstyled"
          fontSize={22}
          color="gray.700"
          _hover={{ color: 'gray.600' }}
          onClick={toggleSettings}
        />
        <Tooltip label="Download data as csv">
          <a
            href={exportSessionAsCsv(sensor.labels, session.data)}
            download={`${sensor.name.toLowerCase()}_${session.id}.csv`}
            target="_blank"
            rel="noreferrer">
            <IconButton
              aria-label="Download exercise data"
              icon={<MdDownload />}
              variant="unstyled"
              fontSize={22}
              color="gray.700"
              _hover={{ color: 'gray.600' }}
            />
          </a>
        </Tooltip>
      </HStack>
    </HStack>
  );
};

export default GraphHeader;
