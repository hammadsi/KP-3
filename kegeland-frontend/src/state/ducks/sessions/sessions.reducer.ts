import { createSlice } from '@reduxjs/toolkit';
import { chain } from 'lodash';

import {
  isFulfilledAction,
  isPendingAction,
  isRejectedAction,
} from '../../../utils/thunk.utils';

import { fetchSessionById, fetchSessions } from './sessions.actions';
import { SessionsState } from './sessions.interface';

export const initialState: SessionsState = {
  loading: false,
  session: undefined,
  data: [],
  error: undefined,
};

const sessionsSlice = createSlice({
  name: 'sessions',
  initialState,
  reducers: {
    clearSessionsState: (state) => {
      state.loading = false;
      state.session = undefined;
      state.data = [];
      state.error = undefined;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchSessionById.fulfilled, (state, action) => {
        const { data, ...payload } = action.payload;
        state.session = {
          ...payload,
          data: chain(data).toPairs().sortBy(0).fromPairs().value(),
        };
      })
      .addCase(fetchSessions.fulfilled, (state, action) => {
        state.data = action.payload;
      })
      .addMatcher(
        (action) => isPendingAction(action, 'sessions'),
        (state) => {
          state.loading = true;
          state.error = undefined;
        },
      )
      .addMatcher(
        (action) => isFulfilledAction(action, 'sessions'),
        (state) => {
          state.loading = false;
          state.error = undefined;
        },
      )
      .addMatcher(
        (action) => isRejectedAction(action, 'sessions'),
        (state, { error }) => {
          state.loading = false;
          state.error = error.message;
        },
      );
  },
});

export const { clearSessionsState } = sessionsSlice.actions;

export default sessionsSlice.reducer;
