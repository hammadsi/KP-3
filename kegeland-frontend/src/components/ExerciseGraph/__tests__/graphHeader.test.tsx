import renderer from 'react-test-renderer';

import { SensorType } from '../../../state/ducks/sensors/sensors.interface';
import GraphHeader from '../GraphHeader';
import { sessionMock } from '../../mocks/sessions.mock';

describe('Test Graph header', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <GraphHeader
          sensor={{
            id: 'sensorId',
            name: SensorType.FEMFIT,
            labels: ['s1', 's2', 's3'],
          }}
          session={sessionMock}
          toggleSettings={() => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
