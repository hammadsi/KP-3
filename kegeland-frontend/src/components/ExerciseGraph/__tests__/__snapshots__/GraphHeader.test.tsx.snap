// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Graph header renders correctly 1`] = `
<div
  className="chakra-stack css-1oeb4ru"
>
  <div
    className="chakra-stack css-he6z7n"
  >
    <div
      className="chakra-stack css-84zodg"
    >
      <div
        className="chakra-stack css-x09s84"
      >
        <p
          className="chakra-text css-1lgbcr5"
        >
          Date of exercise
          :
        </p>
        <div
          className="chakra-stack css-64a7jp"
        >
          <p
            className="chakra-text css-16h254d"
          >
            Jan 1, 1970 1:20 AM
          </p>
        </div>
      </div>
    </div>
    <div
      className="chakra-stack css-84zodg"
    >
      <div
        className="chakra-stack css-x09s84"
      >
        <p
          className="chakra-text css-1lgbcr5"
        >
          Device
          :
        </p>
        <div
          className="chakra-stack css-64a7jp"
        >
          <svg
            className="chakra-icon css-1neie3l"
            fill="currentColor"
            focusable={false}
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M8 23h8v-2H8v2zm8-21.99L8 1c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V3c0-1.1-.9-1.99-2-1.99zM16 15H8V5h8v10z"
            />
          </svg>
          <p
            className="chakra-text css-16h254d"
          >
            Femfit
          </p>
        </div>
      </div>
    </div>
    <div
      className="chakra-stack css-84zodg"
    >
      <div
        className="chakra-stack css-x09s84"
      >
        <p
          className="chakra-text css-1lgbcr5"
        >
          Duration
          :
        </p>
        <div
          className="chakra-stack css-64a7jp"
        >
          <svg
            className="chakra-icon css-1neie3l"
            fill="currentColor"
            focusable={false}
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 2C6.5 2 2 6.5 2 12s4.5 10 10 10 10-4.5 10-10S17.5 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm.5-13H11v6l5.2 3.2.8-1.3-4.5-2.7V7z"
            />
          </svg>
          <p
            className="chakra-text css-16h254d"
          >
            00:00:01
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="chakra-stack css-1sigv9r"
  >
    <button
      aria-label="Open graph settings"
      className="chakra-button css-dlvuv7"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden={true}
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58a.49.49 0 00.12-.61l-1.92-3.32a.488.488 0 00-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54a.484.484 0 00-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58a.49.49 0 00-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
        />
      </svg>
    </button>
    <a
      download="femfit_id.csv"
      href="data:text/csv;charset=utf-8,s1,s2,s3%0A132,1,2,3%0A1233,1,2,3"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onPointerDown={[Function]}
      onPointerEnter={[Function]}
      rel="noreferrer"
      target="_blank"
    >
      <button
        aria-label="Download exercise data"
        className="chakra-button css-dlvuv7"
        type="button"
      >
        <svg
          aria-hidden={true}
          fill="currentColor"
          focusable={false}
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M5 20h14v-2H5v2zM19 9h-4V3H9v6H5l7 7 7-7z"
          />
        </svg>
      </button>
    </a>
  </div>
</div>
`;
